### ISSUE TITLE
Cannot read properties of null (reading 'getVariableForExportName')

### ISSUE THREAD

**Issue by tjx666** [rollup/rollup#4739]
### Rollup Version

2.60.0

### Operating System (or Browser)

MacOS  Ventura 13.0.1 (22A400

### Node Version (if applicable)

16.18.1

### Link To Reproduction

Sorry, that my company project

### Expected Behaviour

Can bundle success

### Actual Behaviour

This issue happen when I upgarde rolup from 2.59.0 to 2.60.0. I can make sure this is come from 2.60.0. If I disable `manualChunks`, it also works. Maybe something wrong in my `manualChunks`

Error Message:

```txt
chunk: effect/utils/index
(!) Mixing named and default exports
https://rollupjs.org/guide/en/#outputexports
The following entry modules are using named and default exports together:
src/binary.js
src/get-image-info.js
src/loader.js
...and 7 other entry modules

Consumers of your bundle will have to use chunk['default'] to access their default export, which may not be what you want. Use `output.exports: 'named'` to disable this warning
[!] TypeError: Cannot read properties of null (reading 'getVariableForExportName')
TypeError: Cannot read properties of null (reading 'getVariableForExportName')
    at getVariableForExportNameRecursive (/Users/yutengjing/code/some-project/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/shared/rollup.js:12408:19)
    at Module.traceVariable (/Users/yutengjing/code/some-project/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/shared/rollup.js:12917:35)
    at Module.getVariableForExportName (/Users/yutengjing/code/some-project/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/shared/rollup.js:12728:35)
    at Module.getRenderedExports (/Users/yutengjing/code/some-project/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/shared/rollup.js:12682:37)
    at Chunk.preRender (/Users/yutengjing/code/some-project/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/shared/rollup.js:14929:64)
    at Bundle.prerenderChunks (/Users/yutengjing/code/some-project/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/shared/rollup.js:16059:19)
    at Bundle.generate (/Users/yutengjing/code/some-project/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/shared/rollup.js:15919:18)
    at /Users/yutengjing/code/some-project/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/shared/rollup.js:23833:27
    at catchUnfinishedHookActions (/Users/yutengjing/code/some-project/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/shared/rollup.js:23247:20)
    at async Promise.all (index 0)
    at build (/Users/yutengjing/code/some-project/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/bin/rollup:1558:5)
    at runRollup (/Users/yutengjing/code/some-project/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/bin/rollup:1668:21)

/Users/yutengjing/code/some-project/packages/editor/utils:
```

rollup config:

```javascript


import fs from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { promisify } from 'node:util';

import { DEFAULT_EXTENSIONS } from '@babel/core';
import { createFilter, dataToEsm } from '@rollup/pluginutils';
import _copy from 'copy';
import _glob from 'glob';
import { paramCase } from 'param-case';
import merge from 'lodash/merge.js';
import uniq from 'lodash/uniq.js';

import postcssPluginInlineSvg from 'postcss-inline-svg';
import postcssPluginUrl from 'postcss-url';

import vitePluginVue2 from '@vitejs/plugin-vue2';
import vitePluginVueJsx from '@vitejs/plugin-vue2-jsx';

import { getBabelInputPlugin as rollupPluginBabel } from '@rollup/plugin-babel';
import rollupPluginCommonjs from '@rollup/plugin-commonjs';
import rollupPluginImport from 'rollup-plugin-import';
import rollupPluginJson from '@rollup/plugin-json';
import rollupPLuginNodeResolve from 'rollup-plugin-node-resolve';
import rollupPluginPostcss from 'rollup-plugin-postcss';
import rollupPluginRequireContext3 from 'rollup-plugin-require-context3';
import rollupPluginTypescript2 from 'rollup-plugin-typescript2';
import rollupPluginUrl from 'rollup-url-plugin';
import rollupPluginWebWorkerLoader from 'rollup-plugin-web-worker-loader';

const glob = promisify(_glob);
const copy = promisify(_copy);

/**
 * @typedef {import('rollup').OutputPlugin} OutputPlugin
 */

const cwdPath = process.cwd();
const assetTypes = `png,svg,jpg,jpeg,webp,gif,mp4,wasm,zip,d\\.ts`;
const assetsRegexp = new RegExp(`\\.(${assetTypes.split(',').join('|')})$`);

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * 文件块名去除路径、后缀
 * /user/xxxx/src/file.vue -> file
 */
export function getChunks({ fileId, srcPath }) {
    // /user/xxx/src/a/b/xxx.png -> a/b/xxx.png
    let targetFile = path.normalize(path.resolve(fileId).replace(path.resolve(srcPath) + '/', ''));
    const setTargetFile = (suffix) => {
        if (targetFile.indexOf(suffix) !== -1) {
            targetFile = targetFile.substring(0, targetFile.indexOf(suffix));
            return targetFile;
        }
    };

    setTargetFile('.vue') || setTargetFile('.ts') || setTargetFile('.js');

    return targetFile;
}

/**
 * 自定义块粒度，打包后会保留目录结构
 */
export function manualChunks({ fileId, srcPath }) {
    if (fileId.includes('rollup-plugin-worker-loader::')) return false;
    if (!fileId.startsWith(path.resolve(srcPath))) {
        return '_chunks_/vendor';
    }
    const chunk = getChunks({
        fileId,
        srcPath,
    });
    console.log('chunk:', chunk);
    return chunk;
}

/**
 * 复制静态资源至打包路径
 * 若开启 importMetaUrl，会在资源同级目录创建 .xx.js 文件，用来支持 import.meta.url 方式引用
 */
export async function copyAssets(buildDest, options) {
    const copyCssPromise = async () => {
        if (options.assets.cssExtractMode === 'chunk') {
            return copy(`${options.srcPath}/**/*.css`, buildDest);
        }
    };

    const copyAssetsPromise = async () => {
        const files = await copy(`${options.srcPath}/**/*.{${assetTypes}}`, buildDest);
        if (options.assets.urlMode === 'importMetaUrl') {
            return Promise.all(
                files.map(async (item) => {
                    const basename = path.basename(item.dest);
                    const code = `export default new URL("./${basename}", import.meta.url).href;`;
                    return fs.writeFile(item.dest + '.js', code);
                }),
            );
        }
    };

    await Promise.all([copyCssPromise(), copyAssetsPromise()]);
    console.log('复制静态资源到: ' + buildDest);
}
export async function copyDTS(buildDest, options) {
    await copy(`${options.srcPath}/**/*.d.ts`, buildDest);
    console.log('复制 d.ts 到：' + buildDest);
}

/**
 * 样式 postcss 打包配置
 * 1. 将样式提取为单一文件
 * 2. chunk 方式打包引入，按需载入（推荐）
 * 3. 内联方式 css in js
 */
export async function createPostcssPlugins(options, dependenciesPattern = []) {
    // 静态资源提取到 cssExtractPath 位置时需调整路径
    const handlerAssetsPath = function (v, cssExtractPath) {
        const dirname = path.resolve(options.srcPath, path.dirname(cssExtractPath));
        const assetRelativeCssPath = path.relative(dirname, v.absolutePath);
        return v.originUrl.startsWith('http') ? v.originUrl : assetRelativeCssPath;
    };

    const getPostcssOptions = (cssExtractPath = false) => ({
        extensions: ['.css', '.less'],
        extract: cssExtractPath,
        plugins: [
            // svg-loader(xxx.svg)
            postcssPluginInlineSvg({
                removeFill: false,
            }),
            // 样式中的静态资源处理
            postcssPluginUrl({
                url:
                    options.assets.urlMode === 'inline'
                        ? 'inline'
                        : (v) => handlerAssetsPath(v, cssExtractPath),
            }),
        ],
        use: [
            [
                'less',
                {
                    relativeUrls: true,
                    javascriptEnabled: true,
                },
            ],
        ],
    });

    const styleFiles = await glob(`${options.srcPath}/**/*.{vue,less}`);

    let plugins = [];
    if (options.assets.cssExtractMode === 'chunk') {
        plugins = styleFiles.map((file) => {
            const targetFile = path.normalize(
                path.resolve(file).replace(path.resolve(options.srcPath) + '/', ''),
            );

            return rollupPluginPostcss({
                ...getPostcssOptions(targetFile + '.css'),
                include: [new RegExp(file)],
            });
        });

        // 兜底 node_modules，全部用 css in js 方式
        plugins.push(
            rollupPluginPostcss({
                ...getPostcssOptions(false),
                include: [/node_modules/],
            }),
        );

        /**
         * 1. 将文件中 import .less 改成 import .css 方式引入
         * 2. 将 .vue 中的 style 标签改成 import .css 方式引入
         */
        plugins.push({
            name: 'rollup-replace-import-css',
            transform(code) {
                const pattern = /import.*('|")([^'"]+\.(less|css))('|")/g;
                if (!pattern.test(code)) return;

                // 获取文件名
                let imports = '';
                code.replace(pattern, (a, b, file) => {
                    const isFromVue = file.includes('.vue?');
                    // 处理 .vue 文件 (比如 .vue?vue&type=style&index=0&lang.css)
                    const basename = isFromVue
                        ? path.basename(
                              file.split('?')[0].replace(path.resolve(options.srcPath) + '/', ''),
                          )
                        : path.normalize(file);

                    // css 方式引入的不做处理
                    if (basename.endsWith('.css')) {
                        return a;
                    }

                    const inNodeModules =
                        file[0] !== '.' &&
                        (file.includes('node_modules') ||
                            dependenciesPattern.some((item) => item.test(file)));

                    // 嵌套依赖不做处理
                    if (inNodeModules) {
                        return a;
                    }

                    imports += `import './${basename}.css';\n`;

                    return a;
                });

                code = imports + code;
                return code;
            },
        });
    } else {
        // 提取为一个文件
        plugins.push(rollupPluginPostcss(getPostcssOptions(options.assets.cssExtractPath)));
    }

    return plugins;
}

/**
 * 最终输出阶段做一些替换逻辑
 * 1. 替换 vpe、antd 等包的 import 地址
 * 2. 替换为 importMetaUrl 文件, xxx.png -> xxx.png.js
 */
export function replaceRenderChunk(format, code, chunk, options) {
    // xxx.png -> xxx.png.js
    const ids = chunk.imports.filter((item) => assetsRegexp.test(item));
    if (ids && options.assets.urlMode === 'importMetaUrl') {
        ids.forEach((id) => {
            const relativePath = path.relative(path.dirname(chunk.facadeModuleId), id);
            code = code.replace(relativePath, relativePath + '.js');
        });
    }

    if (format === 'cjs') {
        // ignore some code
    }

    return code;
}

let cacheWorkerImporters = [];
/**
 * 支持 "web-worker:xxx.js" 或 "xxx.js?worker" 方式引入 web-worker
 * @returns {OutputPlugin[]}
 */
export function createWorkerPlugins() {
    return [
        {
            name: 'rollup-vite-worker',
            resolveId(id, importer, resolveOptions) {
                if (id.includes('?worker')) {
                    id = 'web-worker:' + id.split('?')[0];
                } else {
                    return null;
                }

                return this.resolve(
                    id,
                    importer,
                    Object.assign({ skipSelf: true }, resolveOptions),
                ).then((resolved) => {
                    cacheWorkerImporters.push(resolved?.id || id);
                    return resolved || { id };
                });
            },
        },
        rollupPluginWebWorkerLoader({
            extensions: ['.js', '.ts'],
            // blob url 的方式
            preserveSource: true,
        }),
    ];
}

/**
 * 组件、工具库，按需引用
 * @returns {OutputPlugin[]}
 */
export function createImportPlugins() {
    return [
        // import {Icon} from 'xxx' ->
        // import icon from 'xxx/lib'; import 'xxx/lib/css.js'
        // ignore some code
        // import { set } from 'lodash' -> from 'lodash/set'
        rollupPluginImport({
            libraryName: 'lodash',
            libraryDirectory: '',
            exportName: 'default',
            namedComponent(_name) {
                return _name + '.js';
            },
            include: /\.(vue|js|jsx|ts|tsx)$/,
        }),
    ];
}

/**
 * 支持 "import xxx.html" 或 "import xxx.xsl?raw" 方式引入字符串
 * @returns {OutputPlugin}
 */
export function createRawPlugin() {
    const cacheIds = [];
    const filter = createFilter([/\.html$/]);

    return {
        name: 'rollup-plugin-static-html',
        resolveId(id, importer, resolveOptions) {
            if (!filter(id) && !id.includes('?raw')) return null;
            id = id.split('?')[0];

            return this.resolve(
                id,
                importer,
                Object.assign({ skipSelf: true }, resolveOptions),
            ).then((resolved) => {
                cacheIds.push(resolved?.id || id);
                return resolved || { id };
            });
        },
        transform(source, id) {
            if (!cacheIds.includes(id)) {
                return null;
            }

            return dataToEsm(source);
        },
    };
}

/**
 * 支持将 "xxx.png" 或 "xxx.xsl?url" 变为 base64 地址
 * @returns {OutputPlugin}
 */
export function createInlineUrlPlugin() {
    const getUrlPlugin = (include) =>
        rollupPluginUrl({
            include: include || [
                '**/*.zip',
                '**/*.svg',
                '**/*.png',
                '**/*.jp(e)?g',
                '**/*.gif',
                '**/*.mp4',
                '**/*.webp',
            ],
            // 打包所有
            limit: Infinity,
        });
    const _urlPlugin = getUrlPlugin();

    return Object.assign({}, _urlPlugin, {
        resolveId(id, importer, resolveOptions) {
            if (!id.includes('?url')) {
                return null;
            }

            return this.resolve(
                id.split('?')[0],
                importer,
                Object.assign({ skipSelf: true }, resolveOptions),
            ).then((resolved) => {
                // 继续将标识传递至 load
                resolved.id += '?url';
                return resolved;
            });
        },
        load(id) {
            if (!id.includes('?url')) {
                return _urlPlugin.load(id);
            }

            id = id.split('?')[0];
            return getUrlPlugin(['**/*']).load(id);
        },
    });
}

export async function getBaseRollupConfig(inputFile, options) {
    const pkg = JSON.parse(await fs.readFile(options.packageJsonPath, { encoding: 'utf-8' }));
    const excludeDependencies = []
        .concat(
            (options.external || Object.keys(pkg.dependencies || {})).map((item) => {
                // ignore some code
                return new RegExp('^' + item + '/?');
            }),
            Object.keys(pkg.peerDependencies || {}).map((item) => {
                return new RegExp('^' + item + '/?');
            }),
        )
        .filter((item) => !!item);

    /** @type {import('rollup').RollupOptions} */
    const config = {
        strictDeprecations: true,
        input: inputFile,
        external: (source, importer) => {
            const patterns = []
                .concat(
                    excludeDependencies,
                    // 非内联时忽略静态资源
                    options.assets.urlMode === 'inline' ? null : assetsRegexp,
                    // 以块打包样式时，忽略默认打包
                    options.assets.cssExtractMode === 'chunk' ? /\.css$/ : null,
                )
                .filter((item) => !!item);

            // ?vue&type=style&index=0&lang.css 此类引用必须打包
            if (source.includes('.vue?') && source.includes('&lang.css')) {
                return false;
            }

            if (source.includes('node_modules') || importer?.includes('node_modules')) {
                return false;
            }

            const bool = patterns.some((item) => {
                return item.test(source);
            });

            // web-worker 内的依赖不支持 external
            if (cacheWorkerImporters.some((item) => item.includes(importer))) {
                if (!bool) {
                    // 可能是嵌套依赖
                    cacheWorkerImporters.push(source);
                    cacheWorkerImporters = uniq(cacheWorkerImporters);
                } else {
                    console.warn(
                        `web-worker 无法忽略 "${importer}" 中 "${source}" 的打包，建议使用依赖注入方式传递，请关注包体积`,
                    );
                }

                return false;
            }

            return bool;
        },
        output: [
            {
                format: 'esm',
                dir: 'es',
                manualChunks: options.manualChunks,
                plugins: [
                    {
                        name: 'customRenderChunk',
                        renderChunk: (code, chunk) => options.build.renderChunk('es', code, chunk),
                    },
                ],
            },
            {
                format: 'cjs',
                dir: 'cjs',
                manualChunks: options.manualChunks,
                exports: 'auto',
                plugins: [
                    {
                        name: 'customRenderChunk',
                        renderChunk: (code, chunk) => options.build.renderChunk('cjs', code, chunk),
                    },
                ],
            },
        ],
        plugins: [
            // 资源打包为 base64
            options.assets.urlMode === 'inline' ? createInlineUrlPlugin() : null,
            createRawPlugin(),
            rollupPLuginNodeResolve({
                extensions: ['.mjs', '.js', '.json', '.vue', '.ts', '.tsx'],
            }),
            rollupPluginCommonjs(),
            options.tsconfigPath
                ? rollupPluginTypescript2({
                      tsconfig: options.tsconfigPath,
                      extensions: ['tsx', '.ts', '.js'],
                  })
                : null,
            rollupPluginBabel({
                // runtime 需安装 @babel/runtime
                babelHelpers: 'bundled',
                exclude: '**/node_modules/**',
                configFile: options.babelConfigPath || false,
                extensions: [...DEFAULT_EXTENSIONS, '.ts', '.tsx'],
            }),
            rollupPluginJson(),
            vitePluginVue2({
                isProduction: true,
                style: {
                    preprocessLang: 'less',
                    preprocessOptions: {
                        relativeUrls: true,
                        javascriptEnabled: true,
                    },
                },
            }),
            vitePluginVueJsx(),
            ...(await createPostcssPlugins(options, excludeDependencies)),
            ...createImportPlugins(),
            rollupPluginRequireContext3(),
            ...createWorkerPlugins(),
            ...(options.plugins || []),
        ],
    };

    return config;
}

export default async (options = {}) => {
    options = merge(
        {
            build: {
                // es 打包输出路径
                esDest: path.resolve(cwdPath, 'es'),
                // cjs 打包输出路径
                cjsDest: path.resolve(cwdPath, 'cjs'),

                // 做一些替换工作
                renderChunk: (format, code, chunk) => {
                    return replaceRenderChunk(format, code, chunk, options);
                },
            },
            // 姿态资源处理
            assets: {
                /**
                 * inline: 将静态资源打包成 base64
                 * importMetaUrl: 将 js 代码中静态资源引用自动转为 new URL('./xxx.png', import.meta.url).href 写法
                 * relative: 保持文件相对路径引用 import a from '../xxx.png',
                 */
                urlMode: 'relative',

                /**
                 * extract: 将样式提取为单一文件
                 * chunk: 方式打包引入，按需载入（推荐）
                 * inline: 内联方式 css in js
                 */
                cssExtractMode: 'chunk',

                // cssExtractMode 为 extract 时生效
                cssExtractPath: 'style.css',
            },
            // 模块文件入口
            inputGlob: `${options.srcPath || 'src'}/**/!(*.d|*.spec).{vue,ts,tsx,js,html}`,

            // 源码位置
            srcPath: 'src',
            // ts 配置路径
            tsconfigPath: path.resolve(cwdPath, './tsconfig.json'),
            // babel 配置路径
            babelConfigPath: path.resolve(__dirname, './babel.config.js'),
            // package.json 路径
            packageJsonPath: path.resolve(cwdPath, './package.json'),

            // 自定义分块策略，传 null 可以是单文件打包
            manualChunks: (fileId) => manualChunks({ fileId, srcPath: options.srcPath }),
            external: null,
        },
        options,
    );

    // 入口
    const entryFileList = (await glob(options.inputGlob)).reduce((r, file) => {
        r[getChunks({ fileId: file, srcPath: options.srcPath })] = file;
        return r;
    }, {});

    // 拷贝静态资源
    if (options.assets.urlMode !== 'inline') {
        await copyAssets(options.build.esDest, options);
        await copyAssets(options.build.cjsDest, options);
    } else {
        await copyDTS(options.build.esDest, options);
        await copyDTS(options.build.cjsDest, options);
    }

    const baseConfig = await getBaseRollupConfig(entryFileList, options);
    return [baseConfig];
};
```

**Comment by lukastaegert** [comment]
Without the actual code this is hard to triage. Also, this is far from the current version of Rollup and there have been many changes in the chunking logic between Rollup 2 and 3, which will make any fix for Rollup 2 less likely.

**Comment by tjx666** [comment]
@lukastaegert I install the latest 3.6.0, still occur the same issue.

**Comment by tjx666** [comment]
I finally find it is because one of my deps types field cause:

<img width="1057" alt="image" src="https://user-images.githubusercontent.com/41773861/206182677-146b4521-023b-452a-9e64-0cb8ac8ae285.png">

**Comment by lukastaegert** [comment]
How is it the cause? Can you fix the problem by removing/changing this file?

**Comment by xuyimingwork** [comment]
same problem when using manualChunks

**Comment by areepp** [comment]
In case anyone stumbled upon this, I got this issue when trying to install sentry to my tanstack start app.
Got it fixed by upgrading vite from v7.2.4 to v8.0.0
